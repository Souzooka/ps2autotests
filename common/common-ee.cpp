#include <stdio.h>
#include <stdarg.h>
#include <string.h>
#include <kernel.h>

#define TEST_THREAD_PRIORITY   0x40
#define TEST_THREAD_STACK_SIZE 0x8000

char schedfBuffer[65536];
unsigned int schedfBufferPos = 0;

void schedf(const char *format, ...) {
	va_list args;
	va_start(args, format);
	schedfBufferPos += vsprintf(schedfBuffer + schedfBufferPos, format, args);
	va_end(args);
}

void flushschedf() {
	printf("%s", schedfBuffer);
	schedfBuffer[0] = '\0';
	schedfBufferPos = 0;
}

u32 g_testThreadDoneSema = 0;

extern int test_main(int argc, char *argv[]);

char testThreadStack[TEST_THREAD_STACK_SIZE] __attribute__ ((aligned(16)));
void testThreadProc() {
	//TODO: Print begin/end traces
	//TODO: Pass arguments from main thread to the test function
	test_main(0, NULL);
	SignalSema(g_testThreadDoneSema);
}

int main(int argc, char *argv[]) {
	//A thread is created to ensure the execution environment is the same
	//across all possible boot methods (direct hw boot, ps2link, emulator, etc.)
	
	ee_sema_t semaInfo;
	memset(&semaInfo, 0, sizeof(ee_sema_t));
	semaInfo.init_count = 0;
	semaInfo.max_count  = 1;
	semaInfo.option     = 0;
	g_testThreadDoneSema = CreateSema(&semaInfo);
	
	u32 testThreadId = 0;
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func               = (void*)&testThreadProc;
	threadParam.initial_priority   = TEST_THREAD_PRIORITY;
	threadParam.stack_size         = TEST_THREAD_STACK_SIZE;
	threadParam.stack              = testThreadStack;
	testThreadId = CreateThread(&threadParam);
	
	StartThread(testThreadId, NULL);
	WaitSema(g_testThreadDoneSema);
	
	return 0;
}
