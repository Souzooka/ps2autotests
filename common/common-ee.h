#ifndef PS2AUTOTESTS_COMMON_EE_H
#define PS2AUTOTESTS_COMMON_EE_H

// Defines for MSVC highlighting.  Not intended to actually compile with msc.
#ifdef _MSC_VER
#define __STDC__
#define _EE
#define __attribute__(x)
#endif

#include <stdio.h>
#include <tamtypes.h>

#undef main
#define main test_main

void schedf(const char *format, ...);
void flushschedf();

#endif
