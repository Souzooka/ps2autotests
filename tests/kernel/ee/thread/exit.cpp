#include <common-ee.h>
#include <kernel.h>
#include <string.h>

#define STACK_SIZE 0x8000
#define TEST_THREAD_PRIORITY 0x40

void printResult(u32 threadId, u32 result) {
	if(threadId == result) {
		schedf("returned thread id\n");
	} else {
		schedf("returned %d\n", result);
	}
}

void printThreadStatus(int threadId) {
	ee_thread_status_t threadStat;
	memset(&threadStat, 0, sizeof(ee_thread_status_t));
	int result = ReferThreadStatus(threadId, &threadStat);
	schedf("result: %02x, init prio: %02x current prio: %02x, status: %02x\n", 
		result, threadStat.initial_priority, threadStat.current_priority, threadStat.status);
}

char exitThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void exitThreadProc(u32) {
	ExitThread();
	SleepThread();
}

char exitThreadChangePriorityStack[STACK_SIZE] __attribute__ ((aligned(16)));
void exitThreadChangePriorityProc(u32 newPriority) {
	ChangeThreadPriority(GetThreadId(), newPriority);
	schedf("  stat after changing priority -> ");
	printThreadStatus(GetThreadId());
	ExitThread();
	SleepThread();
}

char exitDeleteThreadStack[STACK_SIZE] __attribute__ ((aligned(16)));
void exitDeleteThreadProc(u32) {
	ExitDeleteThread();
	SleepThread();
}

int createTestThread(void *entry, int prio, void *stack, u32 stackSize) {
	ee_thread_t threadParam;
	memset(&threadParam, 0, sizeof(ee_thread_t));
	threadParam.func             = entry;
	threadParam.initial_priority = prio;
	threadParam.stack            = stack;
	threadParam.stack_size       = stackSize;
	return CreateThread(&threadParam);
}

int main(int argc, char *argv[]) {
	printf("-- TEST BEGIN\n");

	{
		schedf("exit thread:\n");
		
		int threadId = createTestThread((void*)&exitThreadProc, TEST_THREAD_PRIORITY - 0x10, exitThreadStack, STACK_SIZE);
		StartThread(threadId, NULL);
		
		s32 result = StartThread(threadId, NULL);
		schedf("  start after thread exited -> ");
		printResult(threadId, result);

		result = DeleteThread(threadId);
		schedf("  delete after thread exited -> ");
		printResult(threadId, result);

		flushschedf();
	}

	{
		schedf("exit thread (priority change):\n");

		int threadId = createTestThread((void*)&exitThreadChangePriorityProc, TEST_THREAD_PRIORITY - 0x10, exitThreadChangePriorityStack, STACK_SIZE);
		StartThread(threadId, (void*)(TEST_THREAD_PRIORITY - 0x20));
		
		schedf("  stat after thread exited -> ");
		printThreadStatus(threadId);
		
		flushschedf();
	}
	
	{
		schedf("exit delete thread:\n");
		
		int threadId = createTestThread((void*)&exitDeleteThreadProc, TEST_THREAD_PRIORITY - 0x10, exitDeleteThreadStack, STACK_SIZE);
		StartThread(threadId, NULL);

		s32 result = StartThread(threadId, NULL);
		schedf("  start after thread exited -> ");
		printResult(threadId, result);

		result = DeleteThread(threadId);
		schedf("  delete after thread exited -> ");
		printResult(threadId, result);

		flushschedf();
	}
	
	printf("-- TEST END\n");
	
	return 0;
}
