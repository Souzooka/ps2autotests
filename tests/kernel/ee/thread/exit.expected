-- TEST BEGIN
exit thread:
  start after thread exited -> returned thread id
  delete after thread exited -> returned thread id
exit thread (priority change):
  stat after changing priority -> result: 01, init prio: 30, current prio: 20, status: 01
  stat after thread exited -> result: 10, init prio: 30, current prio: 30, status: 10
exit delete thread:
  start after thread exited -> returned -1
  delete after thread exited -> returned -1
-- TEST END
